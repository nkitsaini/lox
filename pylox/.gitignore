.pdm-python
.venv
**/__pycache__
scratch.lox
