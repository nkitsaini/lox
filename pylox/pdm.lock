# This file is @generated by PDM.
# It is not intended for manual editing.

[[package]]
name = "colorama"
version = "0.4.6"
requires_python = "!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*,!=3.6.*,>=2.7"
summary = "Cross-platform colored terminal text."

[[package]]
name = "colored"
version = "1.4.4"
summary = "Simple library for color and formatting to terminal"

[[package]]
name = "exceptiongroup"
version = "1.1.1"
requires_python = ">=3.7"
summary = "Backport of PEP 654 (exception groups)"

[[package]]
name = "iniconfig"
version = "2.0.0"
requires_python = ">=3.7"
summary = "brain-dead simple config-ini parsing"

[[package]]
name = "packaging"
version = "23.1"
requires_python = ">=3.7"
summary = "Core utilities for Python packages"

[[package]]
name = "pluggy"
version = "1.0.0"
requires_python = ">=3.6"
summary = "plugin and hook calling mechanisms for python"

[[package]]
name = "pytest"
version = "7.3.1"
requires_python = ">=3.7"
summary = "pytest: simple powerful testing with Python"
dependencies = [
    "colorama; sys_platform == \"win32\"",
    "exceptiongroup>=1.0.0rc8; python_version < \"3.11\"",
    "iniconfig",
    "packaging",
    "pluggy<2.0,>=0.12",
    "tomli>=1.0.0; python_version < \"3.11\"",
]

[[package]]
name = "pytest-mock"
version = "3.10.0"
requires_python = ">=3.7"
summary = "Thin-wrapper around the mock package for easier use with pytest"
dependencies = [
    "pytest>=5.0",
]

[[package]]
name = "syrupy"
version = "4.0.2"
requires_python = ">=3.8.1,<4"
summary = "Pytest Snapshot Test Utility"
dependencies = [
    "colored<2.0.0,>=1.3.92",
    "pytest<8.0.0,>=7.0.0",
]

[[package]]
name = "tomli"
version = "2.0.1"
requires_python = ">=3.7"
summary = "A lil' TOML parser"

[metadata]
lock_version = "4.2"
groups = ["default", "pytest-mock", "test"]
content_hash = "sha256:ef62d8f8e0f28f900f86fea4f14a0e18fb3026d94874022ba2bcf9b1ef96c910"

[metadata.files]
"colorama 0.4.6" = [
    {url = "https://files.pythonhosted.org/packages/d1/d6/3965ed04c63042e047cb6a3e6ed1a63a35087b6a609aa3a15ed8ac56c221/colorama-0.4.6-py2.py3-none-any.whl", hash = "sha256:4f1d9991f5acc0ca119f9d443620b77f9d6b33703e51011c16baf57afb285fc6"},
    {url = "https://files.pythonhosted.org/packages/d8/53/6f443c9a4a8358a93a6792e2acffb9d9d5cb0a5cfd8802644b7b1c9a02e4/colorama-0.4.6.tar.gz", hash = "sha256:08695f5cb7ed6e0531a20572697297273c47b8cae5a63ffc6d6ed5c201be6e44"},
]
"colored 1.4.4" = [
    {url = "https://files.pythonhosted.org/packages/f3/d6/00203998f27ab30b2417998006ad0608f236740bb129494dd7c5621861e1/colored-1.4.4.tar.gz", hash = "sha256:04ff4d4dd514274fe3b99a21bb52fb96f2688c01e93fba7bef37221e7cb56ce0"},
]
"exceptiongroup 1.1.1" = [
    {url = "https://files.pythonhosted.org/packages/61/97/17ed81b7a8d24d8f69b62c0db37abbd8c0042d4b3fc429c73dab986e7483/exceptiongroup-1.1.1-py3-none-any.whl", hash = "sha256:232c37c63e4f682982c8b6459f33a8981039e5fb8756b2074364e5055c498c9e"},
    {url = "https://files.pythonhosted.org/packages/cc/38/57f14ddc8e8baeddd8993a36fe57ce7b4ba174c35048b9a6d270bb01e833/exceptiongroup-1.1.1.tar.gz", hash = "sha256:d484c3090ba2889ae2928419117447a14daf3c1231d5e30d0aae34f354f01785"},
]
"iniconfig 2.0.0" = [
    {url = "https://files.pythonhosted.org/packages/d7/4b/cbd8e699e64a6f16ca3a8220661b5f83792b3017d0f79807cb8708d33913/iniconfig-2.0.0.tar.gz", hash = "sha256:2d91e135bf72d31a410b17c16da610a82cb55f6b0477d1a902134b24a455b8b3"},
    {url = "https://files.pythonhosted.org/packages/ef/a6/62565a6e1cf69e10f5727360368e451d4b7f58beeac6173dc9db836a5b46/iniconfig-2.0.0-py3-none-any.whl", hash = "sha256:b6a85871a79d2e3b22d2d1b94ac2824226a63c6b741c88f7ae975f18b6778374"},
]
"packaging 23.1" = [
    {url = "https://files.pythonhosted.org/packages/ab/c3/57f0601a2d4fe15de7a553c00adbc901425661bf048f2a22dfc500caf121/packaging-23.1-py3-none-any.whl", hash = "sha256:994793af429502c4ea2ebf6bf664629d07c1a9fe974af92966e4b8d2df7edc61"},
    {url = "https://files.pythonhosted.org/packages/b9/6c/7c6658d258d7971c5eb0d9b69fa9265879ec9a9158031206d47800ae2213/packaging-23.1.tar.gz", hash = "sha256:a392980d2b6cffa644431898be54b0045151319d1e7ec34f0cfed48767dd334f"},
]
"pluggy 1.0.0" = [
    {url = "https://files.pythonhosted.org/packages/9e/01/f38e2ff29715251cf25532b9082a1589ab7e4f571ced434f98d0139336dc/pluggy-1.0.0-py2.py3-none-any.whl", hash = "sha256:74134bbf457f031a36d68416e1509f34bd5ccc019f0bcc952c7b909d06b37bd3"},
    {url = "https://files.pythonhosted.org/packages/a1/16/db2d7de3474b6e37cbb9c008965ee63835bba517e22cdb8c35b5116b5ce1/pluggy-1.0.0.tar.gz", hash = "sha256:4224373bacce55f955a878bf9cfa763c1e360858e330072059e10bad68531159"},
]
"pytest 7.3.1" = [
    {url = "https://files.pythonhosted.org/packages/1b/d1/72df649a705af1e3a09ffe14b0c7d3be1fd730da6b98beb4a2ed26b8a023/pytest-7.3.1-py3-none-any.whl", hash = "sha256:3799fa815351fea3a5e96ac7e503a96fa51cc9942c3753cda7651b93c1cfa362"},
    {url = "https://files.pythonhosted.org/packages/ec/d9/36b65598f3d19d0a14d13dc87ad5fa42869ae53bb7471f619a30eaabc4bf/pytest-7.3.1.tar.gz", hash = "sha256:434afafd78b1d78ed0addf160ad2b77a30d35d4bdf8af234fe621919d9ed15e3"},
]
"pytest-mock 3.10.0" = [
    {url = "https://files.pythonhosted.org/packages/91/84/c951790e199cd54ddbf1021965b62a5415b81193ebdb4f4af2659fd06a73/pytest_mock-3.10.0-py3-none-any.whl", hash = "sha256:f4c973eeae0282963eb293eb173ce91b091a79c1334455acfac9ddee8a1c784b"},
    {url = "https://files.pythonhosted.org/packages/f6/2b/137a7db414aeaf3d753d415a2bc3b90aba8c5f61dff7a7a736d84b2ec60d/pytest-mock-3.10.0.tar.gz", hash = "sha256:fbbdb085ef7c252a326fd8cdcac0aa3b1333d8811f131bdcc701002e1be7ed4f"},
]
"syrupy 4.0.2" = [
    {url = "https://files.pythonhosted.org/packages/35/65/175a5f1ea4811c1ab65b020e40d8cdf9a3cf080837b329ec838765695802/syrupy-4.0.2.tar.gz", hash = "sha256:3c75ab6866580679b2cb9abe78e74c3e2011fffc6333651c6beb2a78a716ab80"},
    {url = "https://files.pythonhosted.org/packages/8a/c8/b0f0fbfbe168ebad5e42aa8a1adb4d90832324164b325f25667dcadda27a/syrupy-4.0.2-py3-none-any.whl", hash = "sha256:dfd1f0fad298eee753de4f2471d4346412c4435885c4b7beea648d4934c6620a"},
]
"tomli 2.0.1" = [
    {url = "https://files.pythonhosted.org/packages/97/75/10a9ebee3fd790d20926a90a2547f0bf78f371b2f13aa822c759680ca7b9/tomli-2.0.1-py3-none-any.whl", hash = "sha256:939de3e7a6161af0c887ef91b7d41a53e7c5a1ca976325f429cb46ea9bc30ecc"},
    {url = "https://files.pythonhosted.org/packages/c0/3f/d7af728f075fb08564c5949a9c95e44352e23dee646869fa104a3b2060a3/tomli-2.0.1.tar.gz", hash = "sha256:de526c12914f0c550d15924c62d72abc48d6fe7364aa87328337a31007fe8a4f"},
]
