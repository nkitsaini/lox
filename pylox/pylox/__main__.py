from . import interpreter

def main():
	interpreter.Runner().start()


if __name__ == "__main__":
	main()