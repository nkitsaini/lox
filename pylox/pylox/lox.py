from typing import TYPE_CHECKING, Any
import colorama

if TYPE_CHECKING:
	from .scanner.tokens import Token
	from .scanner.ast_interpreter import LoxRuntimeError

had_error = False
had_runtime_error = False

def error(line: int, msg: str):
	global had_error
	had_error = True
	
	print(colorama.Fore.RED + "[ERROR] " + colorama.Fore.RESET + f"{colorama.Fore.GREEN}Line {line+1} |{colorama.Fore.RESET} {msg}")

def runtime_error(error: 'LoxRuntimeError'):
	global had_runtime_error
	had_runtime_error = True
	
	print(colorama.Fore.RED + "[RUNTIME ERROR] " + colorama.Fore.RESET + f"{colorama.Fore.GREEN}Line {error.token.line+1} |{colorama.Fore.RESET} {error.args[0]}")

def errorToken(token: 'Token', msg: str):
	global had_error
	had_error = True
	
	line = token.line
	lexem = token.lexeme
	print(colorama.Fore.RED + "[ERROR] " + colorama.Fore.RESET + f"{colorama.Fore.GREEN}Token: {lexem} | Line {line+1} |{colorama.Fore.RESET} {msg}")

def warningToken(token: 'Token', msg: str):
	line = token.line
	lexem = token.lexeme
	print(colorama.Fore.CYAN + "[WARNING] " + colorama.Fore.RESET + f"{colorama.Fore.GREEN}Token: {lexem} | Line {line+1} |{colorama.Fore.RESET} {msg}")
	
# To make easily testable
def lox_print(val: Any):
	print(val)

# To be used in tests
def reset__():
	global had_error
	global had_runtime_error
	had_error = False
	had_runtime_error = False