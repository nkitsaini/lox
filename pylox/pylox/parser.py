def run():
	print("It Works")
