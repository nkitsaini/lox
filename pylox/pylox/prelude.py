# pyright:reportUnusedVariable=warning, reportUnusedImport=none, reportWildcardImportFromLibrary=none
from typing import *
from pprint import pprint
from dataclasses import dataclass, field
from pathlib import Path