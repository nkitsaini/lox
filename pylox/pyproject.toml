[tool.pdm]

[project]
name = "pylox"
version = "0.1.0"
description = ""
authors = [
    {name = "Ankit Saini", email = "nnkitsaini@gmail.com"},
]
dependencies = []
requires-python = ">=3.10"
readme = "README.md"
license = {text = "MIT"}

[tool.pdm.scripts]
start = "python3 -m pylox"

[project.scripts]
pylox = "pylox.__main__:main"

[build-system]
requires = ["pdm-pep517>=1.0"]
build-backend = "pdm.pep517.api"
