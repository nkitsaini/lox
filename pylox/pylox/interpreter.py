from .prelude import *
import argparse
from .scanner.scanner import Scanner
from .scanner.parser import Parser
# from .scanner.ast_printer import AstPrinter
from .scanner.ast_interpreter import AstInterpreter
from . import lox


@dataclass
class Runner:
	interpreter: AstInterpreter = field(default_factory=AstInterpreter)
	def run(self, source: str):
		scanner = Scanner(source)
		tokens = scanner.scan_tokens();
		# print(tokens)
		statements = Parser(tokens).parse()
		if lox.had_error:
			# print("err")
			return
		assert statements is not None
		
		self.interpreter.interpret(statements)

	def run_file(self, file_path: str):
		self.run(Path(file_path).read_text())
		if lox.had_error:
			exit(65)
		if lox.had_runtime_error:
			exit(70)
	
	def run_prompt(self):
		while True:
			try:
				line = input("> ")
				self.run(line)

				lox.had_error = False
			except EOFError:
				break

	def start(self):
		parser = argparse.ArgumentParser()
		parser.add_argument('source', nargs="?", default=None)
		args = parser.parse_args()
		if args.source:
			self.run_file(args.source)
		else:
			self.run_prompt()
