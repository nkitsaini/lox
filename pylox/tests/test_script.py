from pylox import lox
from pylox.interpreter import Runner
from typing import TYPE_CHECKING, Any
import pytest
from pathlib import Path
if TYPE_CHECKING:
	from pytest_mock import MockerFixture

FIXTURE_PATH = Path(__file__).parent/'fixtures'
SCRIPT_PATH = FIXTURE_PATH/'scripts'

@pytest.fixture(autouse=True)
def run_around_tests():
    lox.reset__()

@pytest.mark.parametrize("script", [x.name for x in SCRIPT_PATH.iterdir()])
def test_script(script: str, mocker: 'MockerFixture', snapshot: Any):
	runner = Runner()
	error_spy = mocker.spy(lox, 'error')
	runtime_spy = mocker.spy(lox, 'runtime_error')
	token_spy = mocker.spy(lox, 'errorToken')
	print_spy = mocker.spy(lox, 'lox_print')
	runner.run((SCRIPT_PATH/script).read_text())
	assert error_spy.call_args_list  == snapshot
	assert runtime_spy.call_args_list  == snapshot
	assert token_spy.call_args_list  == snapshot
	assert print_spy.call_args_list  == snapshot
		