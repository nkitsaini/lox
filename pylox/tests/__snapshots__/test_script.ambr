# serializer version: 1
# name: test_clock_script
  _CallList([
  ])
# ---
# name: test_clock_script.1
  _CallList([
  ])
# ---
# name: test_clock_script.2
  _CallList([
  ])
# ---
# name: test_clock_script.3
  _CallList([
    _Call(
      tuple(
        1356998400.0,
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[class.lox]
  _CallList([
  ])
# ---
# name: test_script[class.lox].1
  _CallList([
  ])
# ---
# name: test_script[class.lox].2
  _CallList([
  ])
# ---
# name: test_script[class.lox].3
  _CallList([
    _Call(
      tuple(
        '<Class: MyClass>',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Test 2',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Bill',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Jane',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Jane',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Jane',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        73,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Bob',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        '============================',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Okay, telling aobut me',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        '<function: howTall>',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        99,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'nil',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        99,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Talkative Bob',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'Done',
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[for_loop.lox]
  _CallList([
  ])
# ---
# name: test_script[for_loop.lox].1
  _CallList([
  ])
# ---
# name: test_script[for_loop.lox].2
  _CallList([
  ])
# ---
# name: test_script[for_loop.lox].3
  _CallList([
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        2,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        4,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        5,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        '=================',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        4,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        5,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        '=================',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        2,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        4,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        5,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        '=================',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        '=================',
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[functions.lox]
  _CallList([
  ])
# ---
# name: test_script[functions.lox].1
  _CallList([
  ])
# ---
# name: test_script[functions.lox].2
  _CallList([
  ])
# ---
# name: test_script[functions.lox].3
  _CallList([
    _Call(
      tuple(
        'This is Zero',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'This is non zero',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'This is non zero',
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[functions_anonymous.lox]
  _CallList([
  ])
# ---
# name: test_script[functions_anonymous.lox].1
  _CallList([
  ])
# ---
# name: test_script[functions_anonymous.lox].2
  _CallList([
  ])
# ---
# name: test_script[functions_anonymous.lox].3
  _CallList([
    _Call(
      tuple(
        0,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        4,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        8,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        12,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        16,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        20,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        24,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        28,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        32,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        36,
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[functions_closure.lox]
  _CallList([
  ])
# ---
# name: test_script[functions_closure.lox].1
  _CallList([
  ])
# ---
# name: test_script[functions_closure.lox].2
  _CallList([
  ])
# ---
# name: test_script[functions_closure.lox].3
  _CallList([
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        2,
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[functions_fibonacci_series.lox]
  _CallList([
  ])
# ---
# name: test_script[functions_fibonacci_series.lox].1
  _CallList([
  ])
# ---
# name: test_script[functions_fibonacci_series.lox].2
  _CallList([
  ])
# ---
# name: test_script[functions_fibonacci_series.lox].3
  _CallList([
    _Call(
      tuple(
        0,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        2,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        5,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        8,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        13,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        21,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        34,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        55,
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[if_conditional.lox]
  _CallList([
  ])
# ---
# name: test_script[if_conditional.lox].1
  _CallList([
  ])
# ---
# name: test_script[if_conditional.lox].2
  _CallList([
  ])
# ---
# name: test_script[if_conditional.lox].3
  _CallList([
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        7,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        7,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        7,
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[scoping.lox]
  _CallList([
  ])
# ---
# name: test_script[scoping.lox].1
  _CallList([
  ])
# ---
# name: test_script[scoping.lox].2
  _CallList([
  ])
# ---
# name: test_script[scoping.lox].3
  _CallList([
    _Call(
      tuple(
        'inner a',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'outer b',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'global c',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'outer a',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'outer b',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'global c',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'global a',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'global b',
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        'global c',
      ),
      dict({
      }),
    ),
  ])
# ---
# name: test_script[while_loop.lox]
  _CallList([
  ])
# ---
# name: test_script[while_loop.lox].1
  _CallList([
  ])
# ---
# name: test_script[while_loop.lox].2
  _CallList([
  ])
# ---
# name: test_script[while_loop.lox].3
  _CallList([
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        2,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        3,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        4,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        5,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        1,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        6,
      ),
      dict({
      }),
    ),
    _Call(
      tuple(
        6,
      ),
      dict({
      }),
    ),
  ])
# ---
